lavfi/dnn: Unification of Async and Sync Modes

shubhanshu02 opened the pull request:
### Patch Set Description

This patchset is a part of optional deliverables in the GSoC project [Async Support for TensorFlow Backend in FFmpeg](https://summerofcode.withgoogle.com/projects/#5224576843251712).

- Builds on the patchset from PR #407. 

**Objective**: Unification of async and sync modes in the DNN backends from the filters' perspective.
**Parts under this deliverable**:
    - Unifies the async and sync modes in the DNN backend from the filters' perspective.
    - Adds `TaskItem` based inference in Native backend.
    - Renames `InferenceItem` to `LastLevelTaskItem`.

### Methodology

In this patchset, we plan to deliver a common inference function for async and sync modes to the filters (as compared to the two functions `ff_dnn_execute_model_async` and `ff_dnn_execute_model`). This will simplify the code on the filter side and the filters don't have to care about the mode of execution. With this patchset, the image frames are sent first by the frames and then collected using the `ff_dnn_get_result` function which sends back the input and output frames in a sequential manner.

The execution mode can now be set by using the `async` flag in the `backend_configs` as follows:
- async mode: `backend_configs=async=1`
- sync mode: `backend_configs=async=0`

Another part of this patchset contains the renaming of the `InferenceItem` in the three backends to `LastLevelTaskItem` and the relevant changes to avoid further confusion in the meaning of these structs.

### Patches

2c2a825b067c990f6af78e80316b484b5ae5b1aa Task-based Inference in Native Backend
5db57894287f23894d2db37d2e8745dc8da11dff Unify Execution Modes in DNN Filters
c0b4154e04d2da26116cd00d320183cb2cc6d9e9 Remove synchronous functions from DNN filters
42ce1db6f37eadb130b3426847ee28177e52fdcd Remove Async Flag from DNN Filter Side
5f0a0ed7ed3d11af31f831f237cff916341dd4b8 Rename InferenceItem to LastLevelTaskItem
85fed1e447bd121089acc623999e1a712cc938bd Include dnn_processing in docs of sr and derain filter

guoyejun commented on libavfilter/dnn/dnn_interface.c:
```diff
@@ -33,6 +33,10 @@ DNNModule *ff_get_dnn_module(DNNBackendType backend_type)
     DNNModule *dnn_module;
 
+    // temporary change till unification completes in atleast one backend
+    av_log(NULL, AV_LOG_ERROR, "Execution currently not supported \n");
```
it's not good to disable a feature even it is temporary, we need to refine the patches

guoyejun replied:
it might be a good method for close/internal code, but it is not good for open source.

shubhanshu02 replied:
Okay, got it.
How about reordering the "Task-based Inference in Native Backend" to first and then combining the unification patches in the three backend to one?

guoyejun replied:
looks good, hope the merged patch is not big, :)

shubhanshu02 replied:
I'll try not to make the patch too large. Once refining is complete, I'll tell you. Thank you.

shubhanshu02 replied:
I tried merging the patches. The first two patches are a bit large, but the next two are small.

shubhanshu02 committed:
lavfi/dnn: Task-based Inference in Native Backend

This commit rearranges the code in Native Backend to use the TaskItem
for inference.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
libavfilter: Unify Execution Modes in DNN Filters

This commit unifies the async and sync mode from the DNN filters'
perspective. As of this commit, the Native backend only supports
synchronous execution mode.

Now the user can switch between async and sync mode by using the
'async' option in the backend_configs. The values can be 1 for
async and 0 for sync mode of execution.

This commit affects the following filters:
1. vf_dnn_classify
2. vf_dnn_detect
3. vf_dnn_processing
4. vf_sr
5. vf_derain

This commit also updates the filters vf_dnn_detect and vf_dnn_classify
to send only the input frame and send NULL as output frame instead of
input frame to the DNN backends.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
libavfilter: Remove synchronous functions from DNN filters

This commit removes the unused sync mode specific code from the DNN
filters since the sync and async mode are now unified from the
filters' perspective.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
libavfilter: Remove Async Flag from DNN Filter Side

Remove async flag from filter's perspective after the unification
of async and sync modes in the DNN backend.

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

guoyejun commented on libavfilter/dnn/dnn_backend_native.c:
```diff
@@ -45,9 +45,29 @@ static const AVClass dnn_native_class = {
-                                          int do_ioproc);
+static DNNReturnType execute_model_native(Queue *inference_queue);
+
+static DNNReturnType extract_inference_from_task(TaskItem *task, Queue *inference_queue)
```
fyi, we may change the inference in a new patch after this patch set is finished, it is really confusing.

the struct name can be LastLevelTaskItem, and the function name may extract_lltask_from_task, we use lltask for variable.

The last level task means the task can be passed to the inference engine, so, 'task' for the interface between filter and ffmpeg-dnn, while lltask is the interface between dnn and inference engine.

shubhanshu02 replied:
Sure, got the point. Shall I include this in this patchset as the last patch or include it in the next patch set?

guoyejun replied:
add as the last patch would be better

shubhanshu02 replied:
Okay, I'll add the patch today.

shubhanshu02 replied:
What about the `inference_queue`? Shall I name it `lltask_queue`?

If I understood correctly, the request item should look like this, right?
```c
typedef struct OVRequestItem {
    ie_infer_request_t *infer_request;
    LastLevelTaskItem **lltasks;
    uint32_t lltask_count;
    ie_complete_call_back_t callback;
} OVRequestItem;
```

guoyejun replied:
yes. right.

shubhanshu02 replied:
Okay, thanks.

shubhanshu02 replied:
Added

guoyejun commented on libavfilter/dnn/dnn_backend_native.c:
```diff
@@ -34,6 +34,7 @@
 #define FLAGS AV_OPT_FLAG_FILTERING_PARAM
 static const AVOption dnn_native_options[] = {
     { "conv2d_threads", "threads num for conv2d layer", OFFSET(options.conv2d_threads), AV_OPT_TYPE_INT,  { .i64 = 0 }, INT_MIN, INT_MAX, FLAGS },
+    { "async",          "use DNN async inference",      OFFSET(options.async),          AV_OPT_TYPE_BOOL, { .i64 = 1 },       0,       1, FLAGS },
```
looks that the code in dnn_backend_native.c/h can be a separate patch, as the preparation for the async/sync unification of filter.

shubhanshu02 replied:
like which part? The conversion to `TaskItem` based inference is in a separate patch.

guoyejun replied:
i meant the code in dnn_backend_native.c/h in the second patch. anyway, it doesn't matter to separate it.

shubhanshu02 replied:
Okay

guoyejun commented on libavfilter/dnn/dnn_backend_openvino.c:
```diff
@@ -375,6 +376,11 @@ static DNNReturnType init_model_ov(OVModel *ov_model, const char *input_name, co
         ctx->options.nireq = av_cpu_count() / 2 + 1;
     }
 
+    if (ctx->options.async != 0 && ctx->options.async != 1) {
```
do we need this if? The allowed values are 0 and 1.

    { "async",           "use DNN async inference",       OFFSET(options.async),           AV_OPT_TYPE_BOOL,   { .i64 = 1 },     0,       1, FLAGS },

shubhanshu02 replied:
Oh right. Thanks for the catch.

guoyejun commented on libavfilter/dnn/dnn_backend_openvino.c:
```diff
@@ -724,6 +730,8 @@ DNNModel *ff_dnn_load_model_ov(const char *model_filename, DNNFunctionType func_
     //parse options
     av_opt_set_defaults(ctx);
+    // set default async to 1
+    ctx->options.async = 1;
```
.i64=1 in dnn_backend_common.h  has set the default value as 1.

shubhanshu02 replied:
Oh so this sets the default value. Okay I'll remove this line

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -83,6 +84,7 @@ typedef struct TFRequestItem {
 #define FLAGS AV_OPT_FLAG_FILTERING_PARAM
 static const AVOption dnn_tensorflow_options[] = {
     { "sess_config", "config for SessionOptions", OFFSET(options.sess_config), AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, FLAGS },
+    { "async",  "use DNN async inference",OFFSET(options.async), AV_OPT_TYPE_BOOL, { .i64 = 1 }, 0, 1, FLAGS},
```
need this line?

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -868,6 +870,8 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename, DNNFunctionType func_
     //parse options
     av_opt_set_defaults(ctx);
+    // set default async to 1
+    ctx->options.async = 1;
```
need these 2 lines?

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -883,6 +887,10 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename, DNNFunctionType func_
         ctx->options.nireq = av_cpu_count() / 2 + 1;
     }
 
+    if (ctx->options.async != 0 && ctx->options.async != 1) {
```
these

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -539,5 +539,5 @@ const AVFilter ff_vf_dnn_detect = {
     .outputs       = dnn_detect_outputs,
     .priv_class    = &dnn_detect_class,
-    .activate      = dnn_detect_activate,
+    .activate      = dnn_detect_activate_async,
```
still use dnn_detect_activate here.  It is not good to see 'async' or 'sync' here, since the filter will not care about this concept.  For less change, we may change the function name from dnn_detect_activate_async into dnn_detect_activate_impl, and call dnn_detect_activate_impl in function dnn_detect_activate.  and then delete function  dnn_detect_activate_sync.

```
static int dnn_detect_activate(AVFilterContext *filter_ctx)
{
    DnnDetectContext *ctx = filter_ctx->priv;

    if (ctx->dnnctx.async)  // do not remove this variable, but to add deprecated, refer to 'options' in dnn_filter_common.h
        output warning here to be backward compatible.
   
    return dnn_detect_activate_impl(filter_ctx);
}
```
the same comment for dnn_processing.

guoyejun replied:
ok, please ignore this comment, since I see it is fixed in the next patch.

shubhanshu02 replied:
Sure.

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -891,6 +891,13 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename, DNNFunctionType func_
     }
 
+#if !HAVE_PTHREAD_CANCEL
+    if (ctx->async) {
```
ctx->options.async?

shubhanshu02 replied:
Yes, corrected.

guoyejun commented on libavfilter/dnn_filter_common.h:
```diff
@@ -46,7 +45,6 @@ typedef struct DnnContext {
     { "output",             "output name of the model",   OFFSET(model_outputnames_string), AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, FLAGS },\
     { "backend_configs",    "backend configs",            OFFSET(backend_options),  AV_OPT_TYPE_STRING,    { .str = NULL }, 0, 0, FLAGS },\
     { "options", "backend configs (deprecated, use backend_configs)", OFFSET(backend_options),  AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, FLAGS | AV_OPT_FLAG_DEPRECATED},\
-    { "async",              "use DNN async inference",    OFFSET(async),            AV_OPT_TYPE_BOOL,      { .i64 = 1},     0, 1, FLAGS},
```
do not remove it for backward compatible, just add comment such as: ignored, use backend_configs='async=1' ...

guoyejun replied:
I mean comment updated in the second column, "use DNN async inference"

shubhanshu02 replied:
Like this "use DNN async inference (ignored, use backend_configs='async=1')"?

guoyejun replied:
yes, otherwise, people will see program reports error and ends due to the unknown option, if he used async in the command line.

shubhanshu02 replied:
Done. I had a question - how will people who haven't seen the source code see this comment? Is there any help command for the filter?

guoyejun replied:
yes, it is not complete backward compatible.

ffmpeg -h filter=<name of filter>

shubhanshu02 replied:
I was thinking that we can make it backward compatible by passing this async parameter to the backend load function. 

Since this `async` is present in `DnnContext`, we can pass it to the load model functions and check if the user has set the value there. (in this case, we'd have to set a default value other than 0 or 1)

The only problem with this method is that if the user sets both the values `async=0` and `backend_configs=async=1`, we'll not know which one user was trying to use.

guoyejun replied:
i guess not many people are using the filter option, and such  backward compatible is too complex ...

shubhanshu02 replied:
Okay. I have changed the comment in column 2.

guoyejun commented on libavfilter/dnn/dnn_backend_openvino.c:
```diff
@@ -846,23 +806,46 @@ DNNReturnType ff_dnn_execute_model_async_ov(const DNNModel *model, DNNExecBasePa
+    else {
+        if (model->func_type == DFT_ANALYTICS_CLASSIFY) {
+            // Once we add async support for tensorflow backend and native backend,
+            // we'll combine the two sync/async functions in dnn_interface.h to
```
not catch this point.

we do support asyn for tf backend now.
and sync/async is combined in next patch without any change here.

I know that classification is not supported/tested for sync in the previous, is this the root cause? So we return error here to first keep the same feature as previous.

shubhanshu02 replied:
Yes. I checked with the following two filters in this way and found them to be working with sync mode as well.

```sh
ffmpeg -i cici.jpg -vf dnn_classify=dnn_backend=openvino:model=emotions-recognition-retail-0003.xml:input=data:output=prob_emotion:confidence=0.3:labels=emotions-recognition-retail-0003.label:target=face:backend_configs=async=0,showinfo -f null -
```

```sh
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label:backend_configs=async=0,showinfo -f null -
```

But I did not remove these lines as I wasn't sure if it will work in all cases. Like if there are other cases as well.

guoyejun replied:
the first command line is not correct, dnn_classify will do nothing since no detect result, and the second command line just contain dnn_detect, not dnn_classify.
To use classify, we need to use the two filters together:  -vf dnn_detect=...dnn_classify=...

my concern was about the comment, we may update the comment to be more clear, keeping return error.

shubhanshu02 replied:
Oh yes, I just noticed it didn't return any coordinates of the regions classified.

How about this?
```
        // Classification filter has not been completely tested with the sync
        // mode and on the TensorFlow and Native backends. So, do not support now
```

guoyejun replied:
just the first line is enough.

we can always support here no matter if tf/native backends are supported or not,

shubhanshu02 replied:
Okay. Updated.

guoyejun commented on libavfilter/dnn/dnn_backend_openvino.c:
```diff
@@ -846,23 +806,46 @@ DNNReturnType ff_dnn_execute_model_async_ov(const DNNModel *model, DNNExecBasePa
+        }
+
+        if (ctx->options.batch_size > 1) {
+            avpriv_report_missing_feature(ctx, "batch mode for sync execution");
```
looks that batch mode for sync execution can be supported without much effort, right?  I do not mean you'd add such support or not, just a discussion.

shubhanshu02 replied:
Yes. It seems like we can use the same code we do for async batch mode to support this. Just that we'd need to be sure that the flush frame function is being called for edge cases.

guoyejun replied:
i see

shubhanshu02 replied:
I tried it also and it works fine since all filters which can use `batch_mode` call the flush frame function.

guoyejun replied:
we may enable it in another new patch some time later

shubhanshu02 replied:
Sure

shubhanshu02 committed:
lavfi/dnn: Rename InferenceItem to LastLevelTaskItem

This patch renames the InferenceItem to LastLevelTaskItem in the
three backends to avoid confusion among the meanings of these structs.

The following are the renames done in this patch:

1. extract_inference_from_task -> extract_lltask_from_task
2. InferenceItem -> LastLevelTaskItem
3. inference_queue -> lltask_queue
4. inference -> lltask
5. inference_count -> lltask_count

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

shubhanshu02 committed:
doc/filters.texi: Include dnn_processing in docs of sr and derain filter

Signed-off-by: Shubhanshu Saxena <shubhanshu.e01@gmail.com>

guoyejun commented on libavfilter/dnn/dnn_backend_native.c:
```diff
@@ -185,10 +186,16 @@ DNNModel *ff_dnn_load_model_native(const char *model_filename, DNNFunctionType f
 
     native_model->ctx.class = &dnn_native_class;
     model->options = options;
+    native_model->ctx.options.async = 0;
```
the option has default value

shubhanshu02 replied:
Corrected

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -1156,7 +1156,7 @@ DNNReturnType ff_dnn_execute_model_tf(const DNNModel *model, DNNExecBaseParams *
 
     if (extract_lltask_from_task(task, tf_model->inference_queue) != DNN_SUCCESS) {
-        av_log(ctx, AV_LOG_ERROR, "unable to extract inference from task.\n");
+        av_log(ctx, AV_LOG_ERROR, "unable to extract lltask from task.\n");
```
for log message, we can use the full name: last level task instead of lltask

shubhanshu02 replied:
Sure

guoyejun commented:
the four patches relative to last level task can be merged together, I see it would be large (thanks for considering small patches), it is more nature to be together.

guoyejun commented on libavfilter/dnn/dnn_backend_native.c:
```diff
@@ -425,24 +437,57 @@ DNNReturnType ff_dnn_execute_model_native(const DNNModel *model, DNNExecBasePara
+DNNAsyncStatusType ff_dnn_get_result_native(const DNNModel *model, AVFrame **in, AVFrame **out)
+{
+    NativeModel *native_model = model->model;
+    return ff_dnn_get_async_result_common(native_model->task_queue, in, out);
```
we need to change the function name to ff_dnn_get_result_common, up to you to decide which patch is best for the change.

shubhanshu02 replied:
I merged it in the unification patch.

guoyejun commented:
looks good to me, please send to ffmpeg community. thanks.

@Semmer2  could you also have a test after the patches are sent, thanks.

shubhanshu02 commented:
Sure, I'll send it in a few minutes after rebasing to the latest master. Thank you for reviewing.

guoyejun commented:
lgtm

Semmer2 commented:
Hi,
The new patches may crushed when using tf backend with dnn_detect. It may coused by null frame of dnn_detect_post_proc. Pleas check this.
The command I used is `./ffmpeg -i dog.jpg -vf dnn_detect=dnn_backend=tensorflow:input=image_tensor:output="num_detections&detection_scores&detection_classes&detection_boxes":model=ssd_mobilenet_v2_coco.pb:backend_configs='async=0',showinfo -y face_detect.jpeg`
Please re-check all the dnn filter and backends. :)

shubhanshu02 commented:
I have corrected the patch and tested it with the following commands.

#### Classification 
I tried with these values of async (0,0), (0,1), (1,0), (1,1) where (x,y) = (async value in detect, async value in classify)
```
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:backend_configs='async=0':input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,dnn_classify=dnn_backend=openvino:model=emotions-recognition-retail-0003.xml:input=data:output=prob_emotion:confidence=0.3:labels=emotions-recognition-retail-0003.label:backend_configs='async=0',showinfo -f null -
```
#### Detection
Tried with `async=0` and `async=1`

```
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:backend_configs='async=0':input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,showinfo -f null -
```
```
ffmpeg -i dog.jpg -vf  dnn_detect=dnn_backend=tensorflow:input=image_tensor:output="num_detections&detection_scores&detection_classes&detection_boxes":model=ssd_mobilenet_v2_coco.pb:backend_configs='async=1',showinfo -y face_detect.jpeg
```
#### Derain

Tried with `async=0` and `async=1`

```
ffmpeg -i rain.jpg -vf derain=model=can.pb:dnn_backend=1 derain.jpg -y 
```
```
ffmpeg -i rain.jpg -vf derain=model=can.model:dnn_backend=0 derain.jpg -y 
```

#### SR filter
Tried with `async=0` and `async=1`
```
ffmpeg -i 480p.jpg -vf sr=model=espcn.pb:dnn_backend=1 espcn.jpg -y 
```

```
ffmpeg -i 480p.jpg -vf sr=model=espcn.model:dnn_backend=0 espcn.jpg -y 
```

#### DNN Processing
Tried with `async=0` and `async=1`
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=openvino:model=can.xml:input=x:output=can/sub:backend_configs="input_resizable=1&async=0" -y rain1.jpg
```

```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=tensorflow:model=can.pb:input=x:output=y:backend_configs='async=1' derain.jpg -y
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=native:model=can.model:input=x:output=y:backend_configs='async=0' derain.jpg -y
```

In the native backend, async is not supported, so `async=1` ran in sync mode.

@guoyejun @Semmer2 Sir, can you have a look at the above commands and check if I missed any filter? Thank you.

Semmer2 commented:
Hi Shubhanshu,
LGTM. And the code passed my local test. :)

shubhanshu02 commented:
Thanks. I'll send the updated patches to the mailing list.

guoyejun commented:
we may need to merge the last patch into the patch which causes the issue

shubhanshu02 commented:
> we may need to merge the last patch into the patch which causes the issue

I think the patch `libavfilter: Unify Execution Modes in DNN Filters` is the one that caused this issue. When I build on it, it showed the same bug. The one just before it is related to the native backend, so won't be contributing to this problem.

Shall I merge these? (these =  `libavfilter: Unify Execution Modes in DNN Filters` + `libavfilter: Send only input frame for DNN Detect and Classify`)

guoyejun commented:
> > we may need to merge the last patch into the patch which causes the issue
> 
> I think the patch `libavfilter: Unify Execution Modes in DNN Filters` is the one that caused this issue. When I build on it, it showed the same bug. The one just before it is related to the native backend, so won't be contributing to this problem.
> 
> Shall I merge these? (these = `libavfilter: Unify Execution Modes in DNN Filters` + `libavfilter: Send only input frame for DNN Detect and Classify`)

yes, we'd keep every single patch not breaking things.

guoyejun commented:
i see you just force-pushed the patches, is it a rebase or something new?

shubhanshu02 commented:
> i see you just force-pushed the patches, is it a rebase or something new?

Actually, I saw some CI tests failing, so thought to rebase against the latest master. Nothing new added.

shubhanshu02 commented:
Thank you for reviewing @guoyejun @Semmer2 Sir.
